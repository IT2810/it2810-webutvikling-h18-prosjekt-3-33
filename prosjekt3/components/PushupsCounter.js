import React from 'react';
import Expo from 'expo';
import { StyleSheet, Text, View, TouchableOpacity, Button } from 'react-native';
import * as Storage from './Storage';
import Colors from '../constants/Colors';

export default class PushupsCounter extends React.Component {
  state = {
    pastPushupsCount: 0,
    currentPushupsCount: 0,
    pushupsGoal: 5
  };
  componentDidMount() {
    this.setState({
      pushupsGoal: this.props.pushupsGoal,
    })
    Storage.getGoals().then(goals =>{
      if(goals.hasOwnProperty('pushupsCount')){
        this.setState({pastPushupsCount: goals.pushupsCount})
      }else{
        goals.pushupsCount = 0;
        Storage.storeGoals(goals);
      }
    });
  }
  componentDidUpdate(prevProps, prevState){
    if(this.props.pushupsGoal != prevProps.pushupsGoal){
      this.setState({pushupsGoal: this.props.pushupsGoal})
    }
    if(this.state != prevState){
      Storage.getGoals().then(goals =>{
        goals.pushupsCount = this.state.pastPushupsCount + this.state.currentPushupsCount;
        Storage.storeGoals(goals);
      });
    }
  }

  decrementCounter(totalProgress){
    this.setState({
      currentPushupsCount: (totalProgress-1 < 0)
        ? this.state.currentPushupsCount
        : this.state.currentPushupsCount-1
    })
  }

  incrementCounter(){
    this.setState({currentPushupsCount: this.state.currentPushupsCount+1})
  }

  progressBarStyle(progress, goal){
    let progressWidth = 0;
    let backgroundColor = '#000';
    let progressGoalRelation = progress/goal;
    if(progressGoalRelation>=1){
      progressWidth = 1;
      backgroundColor = '#75CDA8';
    }
    if (progressGoalRelation < 1 && progressGoalRelation > 0.4){
      progressWidth = progressGoalRelation;
      backgroundColor = '#EDBF2C';
    }
    if(progressGoalRelation<=0.4){
      progressWidth = progressGoalRelation;
      backgroundColor = '#DB0200';
    }
    return{
      flex: progressWidth,
      backgroundColor: backgroundColor,
    }
  }

  render() {
    var totalProgress = this.state.pastPushupsCount + this.state.currentPushupsCount;
    var goal = this.state.pushupsGoal;
    return (
      <View style={styles.container}>
        <View style={styles.textContainer}>
          <View style={styles.minusButtonContainer}>
            <Button
              title="-"
              onPress={() => this.decrementCounter(totalProgress)} />
          </View>
          <Text>
            {totalProgress} out of {goal}
          </Text>
          <View style={styles.buttonContainer}>
            <Button
              title="+"
              onPress={() => this.incrementCounter()}
              style={styles.plusButton}/>
          </View>
        </View>
        <View style={styles.progressBar}>
          <View style={this.progressBarStyle(totalProgress, goal)}>
          </View>
        </View>

      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    marginTop: 15,
    alignItems: "center",
    justifyContent: "center"
  },
  progressBar: {
    flex: 1,
    flexDirection: 'row',
    width: 300,
    height: 20,
    backgroundColor: '#eee',
    marginTop: 20,
  },
  textContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: 300,
  },
  plussButton: {
    backgroundColor: '#2db747',
    fontSize: 25,
  },
  buttonContainer: {
    backgroundColor: '#15c129',
    height: 40,
    width: 40,
  },
  minusButtonContainer: {
    backgroundColor: '#c11537',
    height: 40,
    width: 40,
  }
});
