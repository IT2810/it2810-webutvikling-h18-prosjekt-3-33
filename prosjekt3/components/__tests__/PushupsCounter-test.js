import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import PushupsCounter from '../PushupsCounter';


describe('PushupsCounter', () => {

	it('renders correctly', () => {
	  	const pushupCounter = renderer.create(<PushupsCounter />).toJSON();
	  	expect(pushupCounter).toMatchSnapshot();
	});
});