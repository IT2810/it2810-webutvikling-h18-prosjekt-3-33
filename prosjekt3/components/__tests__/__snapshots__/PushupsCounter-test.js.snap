// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PushupsCounter renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "marginTop": 15,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 300,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#c11537",
          "height": 40,
          "width": 40,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            -
          </Text>
        </View>
      </View>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      0
       out of 
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#15c129",
          "height": 40,
          "width": 40,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            +
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#eee",
        "flex": 1,
        "flexDirection": "row",
        "height": 20,
        "marginTop": 20,
        "width": 300,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000",
          "flex": 0,
        }
      }
    />
  </View>
</View>
`;
